source/mir/random/ndvariable.d

```diff
@@ -279,6 +280,105 @@ nothrow @safe version(mir_random_test) unittest
     assert(fabs(x[0] + x[1] + x[2] - 1) < 1e-10);
 }
 
+/++
+Multinomial distribution.
++/
+struct MultinomialVariable(T)
+    if (isFloatingPoint!T)
+{
+    import mir.random.variable : binomialVar;
+
+    ///
+    enum isNdRandomVariable = true;
+    ///
+    alias Element = uint;
+
+    ///
+    const(T)[] probs;
+    size_t N;
+    double norm = 0.0;
+
+
+    /++
+    Params:
+        probs = probabilities of the multinomial distribution
+        N = Number of rolls
+    Constraints: `sum(probs[i]) <= 1`
+    +/
+    this()(size_t N, const(T)[] probs)
+    {
+        this.N = N;
+        this.probs = probs;
+         /// Makes sure probabilities add up to one, by calculating a normalization factor
+        version(assert)
+        {
+            T norm = 0;
+            foreach(k, p; this.probs)
+        {
+            norm += p;
+        }
+            assert(fabs(norm - 1) <= T.epsilon * probs.length * 2);
+        }
+
+    }
+
+    ///
+    pragma(inline, false)
+    void opCall(G)(scope ref G gen, scope size_t[] result)
+        if (isSaturatedRandomEngine!G)
+    {
+        double sum_p = 0.0;
+        uint sum_n = 0;
+
+        foreach(k, p; this.probs)
+        {
+            if (p > 0.0)
+            {
+                auto rv = binomialVar!T(this.N - sum_n, p / (this.norm - sum_p));
+                result[k] = cast(uint)rv(gen);
+
+            }
+            else
+            {
+                result[k] = 0;
+            }
+
+            sum_p += p;
+            sum_n += result[k];
+        }
+
+
+    }
+    /// ditto
+    void opCall(G)(scope G* gen, scope size_t[] result)
```

Comment:
ditto
as was mentioned before it should be scope uint[] result as a user will never use huge arrays for ndvariables.

Reply:
done

Reply:
the overload with pointer engine should be tested as well

Reply:
Agreed. Done.